handle vcalander documents with multiple VEVENTS not in order

eleith opened the pull request:
when importing ics files from third party services (ex: google), long
running recurrence vcalander objects can contain numerous vevents
relating to past exemptions or edits.

these objects are handled incorrectly, leading to incorrect first and/or
last occurence dates from being inferred. as a result, a number of
imported events are not rendered in the calendar.

this PR makes use of EventIterator's constructor that takes in an array
of VEVENTS to properly handle these complex cases.

Also of note is that other parameter signatures of EventIterator's constructor,
will be deprecated, [according to comments in sabre dav EventIterator.php](https://github.com/nextcloud/3rdparty/blob/master/sabre/vobject/lib/Recur/EventIterator.php#L86)

resolves #23069

eleith committed:
handle vcalander documents with recurrence with exemptions or edits

when importing ics files from third party services (ex: google), long
running recurrence vcalander objects can contain numerous vevents
relating to past exemptions or edits.

these objects are handled incorrectly, leading to incorrect first and/or
last occurence dates from being inferred. as a result, a number of
imported events are not rendered in the calendar.

this PR makes use of EventIterator's constructor that takes in an array
of VEVENTS to properly handle these complex cases.

Also of note is that other parameter signatures of EventIterator's constructor,
will be deprecated, according to comments in sabre dav EventIterator.php

Signed-off-by: leith abdulla <online-nextcloud@eleith.com>

eleith commented:
i don't see any dav or caldav tests yet, so i have not added any unit tests

kesselb commented:
> i don't see any dav or caldav tests yet, so i have not added any unit tests

https://www.github.com/nextcloud/server/tree/master/apps%2Fdav%2Ftests%2Funit%2FCalDAV

eleith commented:
> > i don't see any dav or caldav tests yet, so i have not added any unit tests
> 
> https://www.github.com/nextcloud/server/tree/master/apps%2Fdav%2Ftests%2Funit%2FCalDAV

thanks @kesselb , I'll add tests

eleith commented:
added two unit tests based off scenarios from importing a large and recent google calendar export that identified these issues

gjgd reviewed (commented):
Thanks for the PR! I don't know PHP enough to review this so you should ask someone else, but I will note that some lint checks have failed and seem to be blocking.

eleith commented:
> Thanks for the PR! I don't know PHP enough to review this so you should ask someone else, but I will note that some lint checks have failed and seem to be blocking.

thanks @gjgd, i'm fixing the lint issues and adding another suggested reviewer

eleith commented:
fyi, i've been using this patch personally for a month now and with it, i'm able to successfully subscribe to a number of external calendars, all of which contain recurring events.

kesselb commented on apps/dav/lib/CalDAV/CalDavBackend.php:
```diff
@@ -2382,26 +2383,40 @@ protected function addChange($calendarId, $objectUri, $operation, $calendarType
-				$componentType = $component->name;
-				$uid = (string)$component->UID;
-				break;
+			if ($component->name!=='VTIMEZONE') {
```
```suggestion
			if ($component->name !== 'VTIMEZONE') {
```

cc @ChristophWurst seems there is a rule for php-cs-fixer missing.

kesselb commented on apps/dav/lib/CalDAV/CalDavBackend.php:
```diff
@@ -2382,26 +2383,40 @@ protected function addChange($calendarId, $objectUri, $operation, $calendarType
-		if ($componentType === 'VEVENT' && $component->DTSTART) {
-			$firstOccurrence = $component->DTSTART->getDateTime()->getTimeStamp();
+
+		if (count($vEvents) > 0 && $hasDTSTART) {
```
If `$hasDTSTART` is true we have at least one element and `count($vEvents)` is superfluous.

eleith replied:
true, if we assume valid events, since dtstart is required.

eleith committed:
resolve review comments

update styling for conditionals and simplifying conditional by only
relying on dtstart instead of also counting vevents (since they imply
the same condition, assuming valid events)

Signed-off-by: leith abdulla <online-nextcloud@eleith.com>

eleith commented:
fyi: it took my like five tries to get the builds to pass. saw flakiness on mysql tests, user test and login tests.
